sdk/src/user.ts

```diff
@@ -3915,4 +4190,342 @@ export class User {
+			// add open orders IM
+			if (includeOpenOrders) {
+				perpMarginRequirement = perpMarginRequirement.add(
+					new BN(marketPosition.openOrders).mul(OPEN_ORDER_MARGIN_REQUIREMENT)
+				);
+			}
+
+			// weighted unrealized pnl
+			let positionUnrealizedPnl = calculatePositionPNL(
+				market,
+				marketPosition,
+				true,
+				oraclePriceData
+			);
+			let pnlQuotePrice: BN;
+			if (strict && positionUnrealizedPnl.gt(ZERO)) {
+				pnlQuotePrice = BN.min(
+					quoteOraclePriceData.price,
+					quoteSpotMarket.historicalOracleData.lastOraclePriceTwap5Min
+				);
+			} else if (strict && positionUnrealizedPnl.lt(ZERO)) {
+				pnlQuotePrice = BN.max(
+					quoteOraclePriceData.price,
+					quoteSpotMarket.historicalOracleData.lastOraclePriceTwap5Min
+				);
+			} else {
+				pnlQuotePrice = quoteOraclePriceData.price;
+			}
+			positionUnrealizedPnl = positionUnrealizedPnl
+				.mul(pnlQuotePrice)
+				.div(PRICE_PRECISION);
+
+			if (marginCategory !== undefined) {
+				if (positionUnrealizedPnl.gt(ZERO)) {
+					positionUnrealizedPnl = positionUnrealizedPnl
+						.mul(
+							calculateUnrealizedAssetWeight(
+								market,
+								quoteSpotMarket,
+								positionUnrealizedPnl,
+								marginCategory,
+								oraclePriceData
+							)
+						)
+						.div(new BN(SPOT_MARKET_WEIGHT_PRECISION));
+				}
+
+				if (marginBuffer && positionUnrealizedPnl.lt(ZERO)) {
+					positionUnrealizedPnl = positionUnrealizedPnl.add(
+						positionUnrealizedPnl.mul(marginBuffer).div(MARGIN_PRECISION)
+					);
+				}
+			}
+
+			// Add perp contribution: isolated vs cross
+			const isIsolated = this.isPerpPositionIsolated(marketPosition);
+			if (isIsolated) {
+				// derive isolated quote deposit value, mirroring on-chain logic
+				let depositValue = ZERO;
+				// TODO this field(isolatedPositionScaledBalance) should not be undefined in the future, remove ? later
+				if (marketPosition.isolatedPositionScaledBalance?.gt(ZERO)) {
+					const quoteSpotMarket = this.driftClient.getSpotMarketAccount(
+						market.quoteSpotMarketIndex
+					);
+					const quoteOraclePriceData = this.getOracleDataForSpotMarket(
+						market.quoteSpotMarketIndex
+					);
+					const strictQuote = new StrictOraclePrice(
+						quoteOraclePriceData.price,
+						strict
+							? quoteSpotMarket.historicalOracleData.lastOraclePriceTwap5Min
+							: undefined
+					);
+					const quoteTokenAmount = getTokenAmount(
+						marketPosition.isolatedPositionScaledBalance ?? ZERO, //TODO remove ? later
+						quoteSpotMarket,
+						SpotBalanceType.DEPOSIT
+					);
+					depositValue = getStrictTokenValue(
+						quoteTokenAmount,
+						quoteSpotMarket.decimals,
+						strictQuote
+					);
+				}
+				calc.addIsolatedMarginCalculation(
+					market.marketIndex,
+					depositValue,
+					positionUnrealizedPnl,
+					worstCaseLiabilityValue,
+					perpMarginRequirement
+				);
+				calc.addPerpLiability();
+				calc.addPerpLiabilityValue(worstCaseLiabilityValue);
```

Comment:
is this actually used anywhere in sdk? might not wanna add it prematurely if we arent using it

Reply:
no lol it's not! Ya I can delete. Might make things simpler.

Reply:
what about numSpotLiabilities and numPerpLiabilities? Seems like maybe we might not need those either anyway?